ql/instruments/creditdefaultswap.cpp

```diff
@@ -41,23 +42,48 @@ namespace QuantLib {
                                          bool settlesAccrual,
                                          bool paysAtDefaultTime,
                                          const Date& protectionStart,
-                                         const boost::shared_ptr<Claim>& claim)
+                                         const boost::shared_ptr<Claim>& claim,
+                                         const DayCounter& lastPeriodDayCounter,
+                                         const bool rebatesAccrual)
     : side_(side), notional_(notional), upfront_(boost::none),
       runningSpread_(spread), settlesAccrual_(settlesAccrual),
       paysAtDefaultTime_(paysAtDefaultTime), claim_(claim),
       protectionStart_(protectionStart == Null<Date>() ? schedule[0] :
                                                          protectionStart) {
-        QL_REQUIRE(protectionStart_ <= schedule[0],
+        QL_REQUIRE((protectionStart_ <= schedule[0])
+            || (schedule.rule() == DateGeneration::CDS),
                    "protection can not start after accrual");
+
         leg_ = FixedRateLeg(schedule)
             .withNotionals(notional)
             .withCouponRates(spread, dayCounter)
-            .withPaymentAdjustment(convention);
-        upfrontPayment_.reset(new SimpleCashFlow(0.0, schedule[0]));
+            .withPaymentAdjustment(convention)
+            .withLastPeriodDayCounter(lastPeriodDayCounter);
+
+        if(rebatesAccrual) {
+            boost::shared_ptr<FixedRateCoupon> firstCoupon =
+                boost::dynamic_pointer_cast<FixedRateCoupon>(leg_[0]);
+            // adjust to T+3 standard settlement, alternatively add
+            // an arbitrary date to the constructor
+
+            // standard date. Notice that protection start is already adjusted
+            //  to T+1
+            Date rebateDate = schedule.calendar().advance(
+                schedule.calendar().adjust(protectionStart_, convention), 2,
+                Days, convention);
+
+            accrualRebate_.reset(new SimpleCashFlow(
+                firstCoupon->accruedAmount(protectionStart_),
+                rebateDate));
+        }
 
         if (!claim_)
             claim_ = boost::shared_ptr<Claim>(new FaceValueClaim);
         registerWith(claim_);
+        // the upfront payment left intentionally unitialized to indicate
+        //  this CDS has none.
```

Comment:
I'm a bit concerned that this might break user engines, if they wrote ones that rely on the upfront being possibly null but initialized.  Also, as far as I can see, your `MakeCreditDefaultSwap` class also calls the other constructor, so a missing upfront would cause it to be initialized to 0 anyway.

Reply:
Alternatively we could change MakeCreditDefaultSwap to initialize the upfront payment only if withUpfrontRate is provided. I reverted to your suggestion though. In practice it doesn't really matter, cause running only cds don't exist anymore.